File: foundation/stacks-and-queues/duplicate-brackets/duplicate-brackets.java
/* Duplicate Brackets
   Time Complexity: O(2*n) = O(n)
   Space Complexity: O(n)
*/

class Solution{
    
    boolean duplicateBrackets(String str){        
        Stack<Character> st = new Stack<>();
        
        for(int i=0; i<str.length(); i++){
            char ch = str.charAt(i);
            
            if( ch == ')' ){ 
                if( st.peek() == '(' ) // contains no char btw '(' & ')'
                    return true;
                else{
                    while( st.peek() != '(' ){ // remove all elements till ')'
                        st.pop();
                    }
                    st.pop(); // remove '('
                }
            }
            else st.push(ch);
        }
        return false; // means no duplicacy
    }
}

File: foundation/stacks-and-queues/duplicate-brackets/duplicate-brackets.js
/* Duplicate Brackets
   Time Complexity: O(2*n)=O(n)  |  Space Complexity: O(n)
*/

class Solution{
 
    duplicateBrackets(str){   
        let stack = [];  
        
        for(let i=0; i<str.length; i++){        
            let ch = str.charAt(i);
            
            if( ch == ')' ){
                let top = stack.pop();  
                if( top === '(')  // contains no char btw '(' & ')'
                    return true;
                else{
                    while( top !== '('){  // remove all elements till ')'
                        top = stack.pop();
                    }
                }
            }
            else stack.push(ch);
        }
        return false; // means no duplicacy
    }
    
}
